bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ManagedThingProvider.java

```diff
@@ -41,6 +75,159 @@ protected String keyToString(ThingUID key) {
         return key.toString();
     }
 
+    @Override
+    public void onReadyMarkerAdded(ReadyMarker readyMarker) {
+        String bsn = readyMarker.getIdentifier();
+        loadedXmlThingTypes.add(bsn);
+        handleXmlThingTypesLoaded(bsn);
+        logger.debug("marker added {}", readyMarker);
+    }
+
+    private void handleXmlThingTypesLoaded(String bsn) {
+        logger.debug("handleXmlThingTypesLoaded({})", bsn);
+        thingHandlerFactories.stream().filter(thingHandlerFactory -> bsn.equals(getBundleName(thingHandlerFactory)))
+                .forEach(thingHandlerFactory -> thingHandlerFactoryAdded(thingHandlerFactory));
+    }
+
+    private void thingHandlerFactoryAdded(ThingHandlerFactory thingHandlerFactory) {
+        logger.debug("thingHandlerFactoryAdded({})", thingHandlerFactory.getClass().getName());
+        Collection<Thing> storageList = super.getAll();
+        logger.debug("storageList.size(): {}", storageList.size());
+        storageList.stream().forEach(thing -> createThingFromStorageForThingHandlerFactory(thing, thingHandlerFactory));
+    }
+
+    private boolean compareChannels(Channel a, Channel b) {
+        if (!a.getUID().equals(b.getUID())) {
+            return false;
+        }
+
+        ChannelTypeUID aChannelTypeUID = a.getChannelTypeUID();
+        ChannelTypeUID bChannelTypeUID = b.getChannelTypeUID();
+        if (aChannelTypeUID != null) {
+            if (!aChannelTypeUID.equals(bChannelTypeUID)) {
+                return false;
+            }
+        } else if (bChannelTypeUID != null) {
+            return false;
+        }
+
+        String aAcceptedItemType = a.getAcceptedItemType();
+        String bAcceptedItemType = b.getAcceptedItemType();
+
+        if (aAcceptedItemType != null) {
+            if (!aAcceptedItemType.equals(bAcceptedItemType)) {
+                return false;
+            }
+        } else if (bAcceptedItemType != null) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private void createThingFromStorageForThingHandlerFactory(Thing storedThing, ThingHandlerFactory factory) {
+        logger.debug("createThingFromStorageForThingHandlerFactory({}, {}", storedThing.getUID(),
+                factory.getClass().getName());
+        if (!loadedXmlThingTypes.contains(getBundleName(factory))) {
+            return;
+        }
+
+        logger.info("*** thing.getHandler() = {}", storedThing.getHandler());
```

Comment:
This logger revealed the main issue with this approach: Since the order of ReadyMarker-Calls is not guaranteed, the following could happen, that the RadyMarker of the ThingManager fires first, then the handler is assigned and initalized and everything is fine here. But it could also happen, that the ReadyMarker here fires first. Then we do not have a handler (yet). So any update of the thing would not cause the handler to update the thing, as there is no handler. This ends up in an inconsistent model, where `thing.getHandler() != thing` inside the eventually assigned handler.

Reply:
What about queuing that for later re-try? The GenericThingProvider also queues thing creation when not all XMLs are loaded. So if `handler==null` Add to queue and try if the handler is available let‘s say every 500ms or so.

Reply:
I was also thinking about this but first searhed for a simpler solution.